h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,701 @@
+    * @returns {Object} The full group object associated with the annotation.
+    */
+  vm.group = function() {
+    return groups.get(model.group);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#tagsAutoComplete.
+    * @returns {Promise} immediately resolved to {string[]} -
+    * the tags to show in autocomplete.
+    */
+  vm.tagsAutoComplete = function(query) {
+    return $q.when(tags.filter(query));
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#isHighlight.
+    * @returns {boolean} True if the annotation is a highlight.
+    */
+  vm.isHighlight = function() {
+    var targetLength = (model.target || []).length;
+    var referencesLength = (model.references || []).length;
+    var tagsLength = (model.tags || []).length;
+    return (targetLength && !referencesLength && !(model.text || tagsLength));
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#isPrivate
+    * @returns {boolean} True if the annotation is private to the current user.
+    */
+  vm.isPrivate = function() {
+    return permissions.isPrivate(vm.annotation.permissions, model.user);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#isShared
+    * @returns {boolean} True if the annotation is shared (either with the
+    * current group or with everyone).
+    */
+  vm.isShared = function() {
+    return permissions.isShared(vm.annotation.permissions, model.group);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#setPrivacy
+    *
+    * Set the privacy settings on the annotation to a predefined
+    * level. The supported levels are 'private' which makes the annotation
+    * visible only to its creator and 'shared' which makes the annotation
+    * visible to everyone in the group.
+    *
+    * The changes take effect when the annotation is saved
+    */
+  vm.setPrivacy = function(privacy) {
+    // When the user changes the privacy level of an annotation they're
+    // creating or editing, we cache that and use the same privacy level the
+    // next time they create an annotation.
+    // But _don't_ cache it when they change the privacy level of a reply.
+    if (!model.references) {  // If the annotation is not a reply.
+      permissions.setDefault(privacy);
+    }
+    if (privacy === 'private') {
+      vm.annotation.permissions = permissions.private();
+    } else if (privacy === 'shared') {
+      vm.annotation.permissions = permissions.shared(model.group);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotaitonController#hasContent
+    * @returns {boolean} True if the currently edited annotation has
+    *          content (ie. is not just a highlight)
+    */
+  vm.hasContent = function() {
+    var textLength = (vm.annotation.text || '').length;
+    var tagsLength = (vm.annotation.tags || []).length;
+    return (textLength > 0 || tagsLength > 0);
```

Comment:
Skimming this now I can see that the code is inconsistent about whether it relies on implicit conversions from numbers to booleans. Despite what I said about `if (<expr>)` being idiomatic JS, I still think it is preferable to use explicit tests for `number > 0` rather than relying on 0 being falsey.

Don't need to make changes now though.